File: desktop/windows/src/renderer/src/lib/parityContracts.test.ts
// Windows conformance suite for the shared cross-platform parity contracts
// (contracts/parity/README.md). Runs the repo-root fixture vectors through the
// REAL production rules: task due-date bucketing (taskBuckets.bucketOf, the
// fold_overdue model shared with macOS), local-day conversation grouping
// (filtering.startOfLocalDay / groupConversationsByDate), and relative day
// labels (formatDue badge + conversation section headers), including the DST
// transition-day cases that only execute on runners whose zone observes the
// transition (counted skips elsewhere; the non-DST cases always run).
import { readFileSync } from 'node:fs'
import { fileURLToPath } from 'node:url'
import { describe, expect, it } from 'vitest'
import { bucketOf, formatDue, type Bucket } from './taskBuckets'
import { groupConversationsByDate, startOfLocalDay } from './conversations/filtering'
import type { ConversationRow } from './pageCache'

function fixture<T>(name: string): T {
  const path = fileURLToPath(new URL(`../../../../../../contracts/parity/${name}`, import.meta.url))
  return JSON.parse(readFileSync(path, 'utf8')) as T
}

/** Fixture local-time components [year, month, day, hour?, minute?] → epoch ms
 *  in the runner's zone (the contracts are calendar rules, zone-independent). */
type Components = [number, number, number, number?, number?]
const localMs = (c: Components): number => new Date(c[0], c[1] - 1, c[2], c[3] ?? 0, c[4] ?? 0).getTime()

const RELATIVE_LABELS = ['Today', 'Tomorrow', 'Yesterday']

describe('task due buckets (fold_overdue model)', () => {
  type BucketCase = {
    name: string
    now: Components
    due: Components | null
    created: Components | null
    expected: { fold_overdue: string; separate_overdue: string }
  }
  // Fixture bucket names → this platform's Bucket vocabulary. fold_overdue never
  // expects 'overdue'; an unmapped name fails loudly rather than passing vacuously.
  const BUCKET_MAP: Record<string, Bucket> = {
    today: 'today',
    tomorrow: 'tomorrow',
    later: 'later',
    no_deadline: 'nodate'
  }
  const { cases } = fixture<{ cases: BucketCase[] }>('task_due_buckets.json')
  it.each(cases)('$name', (c) => {
    const now = localMs(c.now)
    const dueAt = c.due === null ? null : localMs(c.due)
    expect(bucketOf({ dueAt }, now)).toBe(BUCKET_MAP[c.expected.fold_overdue])
  })
})

describe('local day keys', () => {
  type DayKeyCase = { name: string; utc: string; expected_by_offset: Record<string, string> }
  const pad = (n: number): string => String(n).padStart(2, '0')
  const { cases } = fixture<{ cases: DayKeyCase[] }>('day_keys.json')
  for (const c of cases) {
    const instant = Date.parse(c.utc)
    const offsetEast = -new Date(instant).getTimezoneOffset()
    const expected = c.expected_by_offset[String(offsetEast)]
    // Only cases whose fixture covers this runner's zone offset execute; offset 0
    // is always present so UTC CI runs every case.
    it.runIf(expected !== undefined)(`${c.name} (offset ${offsetEast})`, () => {
      const d = new Date(startOfLocalDay(instant))
      expect(`${d.getFullYear()}-${pad(d.getMonth() + 1)}-${pad(d.getDate())}`).toBe(expected)
    })
  }
})

describe('relative day labels', () => {
  type LabelCase = {
    name: string
    now: Components
    target_day: Components
    requires_dst_transition_between?: [Components, Components]
    labels: { conversation_section: string | null; due_badge: string | null }
  }
  const transitionObserved = (c: LabelCase): boolean => {
    if (!c.requires_dst_transition_between) return true
    const [a, b] = c.requires_dst_transition_between
    return (
      new Date(a[0], a[1] - 1, a[2]).getTimezoneOffset() !== new Date(b[0], b[1] - 1, b[2]).getTimezoneOffset()
    )
  }
  const { cases } = fixture<{ cases: LabelCase[] }>('section_labels.json')
  for (const c of cases) {
    it.runIf(transitionObserved(c))(c.name, () => {
      const now = localMs(c.now)
      const targetNoon = localMs([c.target_day[0], c.target_day[1], c.target_day[2], 12, 0])

      const badge = formatDue(targetNoon, now)
      if (c.labels.due_badge === null) {
        expect(badge).toBeTruthy()
        expect(RELATIVE_LABELS).not.toContain(badge)
      } else {
        expect(badge).toBe(c.labels.due_badge)
      }

      const row = {
        id: 'parity-row',
        title: 'row',
        subtitle: '',
        preview: '',
        source: 'cloud',
        sortAt: targetNoon
      } as unknown as ConversationRow
      const sections = groupConversationsByDate([row], now)
      expect(sections).toHaveLength(1)
      const label = sections[0].label
      if (c.labels.conversation_section === null) {
        expect(label).toBeTruthy()
        expect(RELATIVE_LABELS).not.toContain(label)
      } else {
        expect(label).toBe(c.labels.conversation_section)
      }
    })
  }
})

File: app/test/parity/parity_contracts_test.dart
import 'dart:convert';
import 'dart:io';

import 'package:flutter_test/flutter_test.dart';

import 'package:omi/backend/schema/gen/action_items_folders_wire.g.dart';
import 'package:omi/pages/action_items/task_categorization.dart';
import 'package:omi/providers/conversation_provider.dart';

/// Flutter conformance suite for the shared cross-platform parity contracts
/// (contracts/parity/README.md). Runs the repo-root fixture vectors through the
/// REAL production rules: task bucketing (categorizeTasks, this platform's
/// separate_overdue model), local-day conversation keys
/// (conversationLocalDayKey, the #10198 contract), and action item wire decode
/// (GeneratedActionItemResponse.fromJson). Day-key cases execute only when the
/// fixture covers the runner's zone offset at that instant; offset 0 is always
/// present so UTC CI runs every case.
void main() {
  final root = _repoRoot();

  group('task due buckets (separate_overdue model)', () {
    final fixture = _fixture(root, 'task_due_buckets.json');
    const bucketByName = {
      'today': TaskCategory.today,
      'tomorrow': TaskCategory.tomorrow,
      'later': TaskCategory.later,
      'no_deadline': TaskCategory.noDeadline,
      'overdue': TaskCategory.overdue,
    };
    for (final raw in fixture['cases'] as List<dynamic>) {
      final c = raw as Map<String, dynamic>;
      test(c['name'] as String, () {
        final now = _localFromComponents(c['now'] as List<dynamic>);
        final due = c['due'] == null ? null : _localFromComponents(c['due'] as List<dynamic>);
        final created = c['created'] == null ? null : _localFromComponents(c['created'] as List<dynamic>);
        final item = _wireItem(due: due, created: created);

        final categorized = categorizeTasks([item], false, now: now);
        final actual = categorized.entries.where((e) => e.value.contains(item)).map((e) => e.key).toList();
        final expectedName = (c['expected'] as Map<String, dynamic>)['separate_overdue'] as String;

        expect(actual, [bucketByName[expectedName]!]);
      });
    }
  });

  group('local day keys', () {
    final fixture = _fixture(root, 'day_keys.json');
    for (final raw in fixture['cases'] as List<dynamic>) {
      final c = raw as Map<String, dynamic>;
      final instant = DateTime.parse(c['utc'] as String);
      final offsetEast = instant.toLocal().timeZoneOffset.inMinutes;
      final expected = (c['expected_by_offset'] as Map<String, dynamic>)['$offsetEast'] as String?;
      test('${c['name']} (offset $offsetEast)', () {
        final key = conversationLocalDayKey(instant);
        expect('${key.year}-${_pad(key.month)}-${_pad(key.day)}', expected);
      }, skip: expected == null ? 'fixture does not cover this zone offset' : false);
    }
  });

  group('action item wire decode (parity contract)', () {
    final fixture = _fixture(root, 'wire_action_item.json');
    for (final raw in fixture['cases'] as List<dynamic>) {
      final c = raw as Map<String, dynamic>;
      test(c['name'] as String, () {
        final byModel = c['expected_by_model'] as Map<String, dynamic>?;
        if (byModel != null) {
          // This client is the strict_decode model: a present-but-unparseable
          // due_at rejects the whole item (see the README divergence register).
          final strict = byModel['strict_decode'] as Map<String, dynamic>;
          expect(strict['parses'], isFalse);
          expect(
            () => GeneratedActionItemResponse.fromJson(c['payload'] as Map<String, dynamic>),
            throwsFormatException,
          );
          return;
        }
        final expected = c['expected'] as Map<String, dynamic>;
        final item = GeneratedActionItemResponse.fromJson(c['payload'] as Map<String, dynamic>);

        expect(expected['parses'], isTrue);
        expect(item.description, expected['description']);
        expect(item.completed, expected['completed']);
        final dueUtc = expected['due_utc'] as String?;
        if (dueUtc == null) {
          expect(item.dueAt, isNull);
        } else {
          expect(item.dueAt?.millisecondsSinceEpoch, DateTime.parse(dueUtc).millisecondsSinceEpoch);
        }
      });
    }
  });
}

/// Walk up from the package dir to the repo root (the dir holding
/// contracts/parity), so the suite works from either the app dir or repo root.
Directory _repoRoot() {
  var dir = Directory.current.absolute;
  for (var i = 0; i < 6; i++) {
    if (Directory('${dir.path}${Platform.pathSeparator}contracts${Platform.pathSeparator}parity').existsSync()) {
      return dir;
    }
    final parent = dir.parent;
    if (parent.path == dir.path) break;
    dir = parent;
  }
  throw StateError('contracts/parity not found above ${Directory.current.path}');
}

Map<String, dynamic> _fixture(Directory root, String name) {
  final file = File(
    '${root.path}${Platform.pathSeparator}contracts${Platform.pathSeparator}parity'
    '${Platform.pathSeparator}$name',
  );
  return jsonDecode(file.readAsStringSync()) as Map<String, dynamic>;
}

/// Fixture local-time components [year, month, day, hour?, minute?] in the
/// runner's zone (the contracts are calendar rules, zone-independent).
DateTime _localFromComponents(List<dynamic> c) =>
    DateTime(c[0] as int, c[1] as int, c[2] as int, c.length > 3 ? c[3] as int : 0, c.length > 4 ? c[4] as int : 0);

String _pad(int n) => n.toString().padLeft(2, '0');

/// Build the item through the production wire decode so bucket cases exercise
/// the same path a backend response takes.
GeneratedActionItemResponse _wireItem({DateTime? due, DateTime? created}) => GeneratedActionItemResponse.fromJson({
      'id': 'parity',
      'description': 'parity case',
      'completed': false,
      if (created != null) 'created_at': created.toUtc().toIso8601String(),
      if (due != null) 'due_at': due.toUtc().toIso8601String(),
    });
